packages/@aws-cdk/aws-s3/lib/notifications-resource/notifications-resource.ts

```diff
@@ -0,0 +1,165 @@
+
+    constructor(parent: cdk.Construct, id: string, props: NotificationsProps) {
+        super(parent, id);
+        this.bucket = props.bucket;
+    }
+
+    /**
+     * Adds a notification subscription for this bucket.
+     * If this is the first notification, a BucketNotification resource is added to the stack.
+     *
+     * @param event The type of event
+     * @param target The target construct
+     * @param s3KeyFilters A set of S3 key filters
+     */
+    public addNotification(event: EventType, target: INotificationDestination, ...s3KeyFilters: string[]) {
+        this.createResourceOnce();
+
+        // resolve target. this also provides an opportunity for the target to e.g. update
+        // policies to allow this notification to happen.
+        const targetProps = target.bucketNotificationDestination(this.bucket);
+        const commonConfig: CommonConfiguration = {
+            Events: [ event ],
+            Filter: renderFilter(s3KeyFilters),
+        };
+
+        // based on the target type, add the the correct configurations array
+        switch (targetProps.type) {
+            case NotificationDestinationType.Lambda:
+                this.lambdaNotifications.push({ ...commonConfig, LambdaFunctionArn: targetProps.arn });
+                break;
+
+            case NotificationDestinationType.Queue:
+                this.queueNotifications.push({ ...commonConfig, QueueArn: targetProps.arn });
+                break;
+
+            case NotificationDestinationType.Topic:
+                this.topicNotifications.push({ ...commonConfig, TopicArn: targetProps.arn });
+                break;
+
+            default:
+                throw new Error('Unsupported notification target type:' + NotificationDestinationType[targetProps.type]);
+        }
+    }
+
+    private renderNotificationConfiguration(): NotificationConfiguration {
+        return {
+            LambdaFunctionConfigurations: this.lambdaNotifications.length > 0 ? this.lambdaNotifications : undefined,
+            QueueConfigurations: this.queueNotifications.length > 0 ? this.queueNotifications : undefined,
+            TopicConfigurations: this.topicNotifications.length > 0 ? this.topicNotifications : undefined
+        };
+    }
+
+    /**
+     * Defines the bucket notifications resources in the stack only once.
+     * This is called lazily as we add notifications, so that if notifications are not added,
+     * there is no notifications resource.
+     */
+    private createResourceOnce() {
+        if (!this.customResourceCreated) {
+            const handlerArn = NotificationsResourceHandler.singleton(this);
+            new cdk.Resource(this, 'Resource', {
+                type: 'Custom::S3BucketNotifications',
+                properties: {
+                    ServiceToken: handlerArn,
+                    BucketName: this.bucket.bucketName,
+                    NotificationConfiguration: new cdk.Token(() => this.renderNotificationConfiguration())
```

Comment:
Is it okay for the value of this thing to be an empty object? Or do we have to return `undefined` in that case to prevent the key from being added?